### ISSUE TITLE
Complaints when calling toString() of an IndexOutOfBoundsException

### ISSUE THREAD

**Issue by slyubomirsky** [OpenJML/OpenJML#630]
JML gives an error in the following code, caliming it cannot establish the precondition `\typeof(this) == \type(Throwable)` for `th` below (or the same if `ex` is used directly):

```
    static class XYZ {
        //@ private exceptional_behavior
        //@     signals (IndexOutOfBoundsException ex) true;
        private static void callMe() {
            throw new IndexOutOfBoundsException();
        }
        //@ private behavior
        //@     ensures true;
        private static boolean convertException() {
            try {
                callMe();
            } catch (IndexOutOfBoundsException ex) {
                //@ assert ex.toStringDefined;
                Object obj = ex;
                String s = obj.toString();
                Throwable th = ex;
                s = th.toString();
                throw new ArrayIndexOutOfBoundsException(s);
            }
            return true;
        }
    }
```

However, the verification goes through if the exception is cast directly to object, as here:

```
    static class XYZ {
        //@ private exceptional_behavior
        //@     signals (IndexOutOfBoundsException ex) true;
        private static void callMe() {
            throw new IndexOutOfBoundsException();
        }
        //@ private behavior
        //@     ensures true;
        private static boolean convertException() {
            try {
                callMe();
            } catch (IndexOutOfBoundsException ex) {
                throw new ArrayIndexOutOfBoundsException(((Object)ex).toString());
            }
            return true;
        }
    }
```

**Comment by RustanLeino** [comment]
Here is one further test case. This one shows that the bug is not specific to `Throwable`. However, amazingly, it seems that "`toString`" must be a substring of the method name. For other method names, the verifier works as expected.

        static class MySuperclass {
            /*@ public normal_behavior
              @     requires true;
              @*/
            public void XYtoStringx() { } // rename this method (and the override and the call below) to, for example, XYtoXtringx and the bug is not triggered
        }
        static class MySubclass extends MySuperclass {
            /*@ also
              @ public normal_behavior
              @     requires false;
              @*/
            public void XYtoStringx() { }
        }
        static void testMethod(MySubclass sub) {
            sub.XYtoStringx();  // this complains that the precondition false in MySubclass is not satisfied, but it shouldn't complain about that
        }

  Rustan

**Comment by davidcok** [comment]
Yes - sorry guys - some wayward debugging code for exploring some unusual situations with toString() inheritance. That problem was fixed another way, but two lines were forgotten. I’m testing a fix and will push it to ‘development’ tomorrow if the tests don’t surface any problems. Thanks for alerting me and sorry for the work to create test cases. (And yes they were conditioned on the string “toString” as Rustan discovered).

- David

**Comment by davidcok** [comment]
Pushed to the HEAD of development.
